megamek/src/megamek/client/ratgenerator/ForceDescriptor.java

```diff
@@ -0,0 +1,1696 @@
+     *                     could not be generated with the given parameters.
+	 */
+	private List<ModelRecord> generateNovaFormation(List<ForceDescriptor> subs, int networkMask, int numGroups) {
+	    //Split base and infantry units
+	    List<ForceDescriptor> baseSubs = new ArrayList<>();
+        List<ForceDescriptor> baSubs = new ArrayList<>();
+        List<ForceDescriptor> infSubs = new ArrayList<>();
+        for (ForceDescriptor sub : subs) {
+            if (sub.getUnitType() == UnitType.BATTLE_ARMOR) {
+                baSubs.add(sub);
+            } else if (sub.getUnitType() == UnitType.INFANTRY) {
+                infSubs.add(sub);
+            } else {
+                baseSubs.add(sub);
+            }
+        }
+        //If there is any conventional infantry we'll generate it first, then assign the APC role
+        //to as many vehicles (if any) in the base units as we have foot infantry. Any remaining vehicles
+        //will get the infantry support role.
+        if (infSubs.size() > 0) {
+            generateLance(infSubs);
+            int footCount = (int)infSubs.stream().filter(fd -> fd.getMovementModes()
+                    .contains(EntityMovementMode.INF_LEG)).count();
+            for (int i = 0; i < baseSubs.size(); i++) {
+                if (baseSubs.get(i).getUnitType() == UnitType.TANK
+                        || baseSubs.get(i).getUnitType() == UnitType.VTOL) {
+                    if (footCount > 0) {
+                        baseSubs.get(i).getRoles().add(MissionRole.APC);
+                        footCount--;
+                    } else {
+                        baseSubs.get(i).getRoles().add(MissionRole.INF_SUPPORT);
+                    }
+                }
+            }
+        }
+        //Generate the base units according to the formation type.
+        List<ModelRecord> baseUnitList = null;
+        if (!baseSubs.isEmpty()) {
+            baseUnitList = generateFormation(baseSubs, networkMask, numGroups);
+        }
+        if (null == baseUnitList) {
+            generateLance(baseSubs);
+            baseUnitList = baseSubs.stream().map(ForceDescriptor::getModelName)
+                    .map(m -> RATGenerator.getInstance().getModelRecord(m))
+                    .filter(Objects::nonNull).collect(Collectors.toList());
+        }
+
+        //Any BA in exceess of the number of omni base units will require mag clamps, up to the number of base units.
+        int magReq = Math.min((int)(baSubs.size() - baseUnitList.stream().filter(mr -> mr.isOmni()).count()),
+                baSubs.size());
+        for (int i = 0; i < baSubs.size(); i++) {
+            if (i < magReq) {
+                baSubs.get(i).getRoles().add(MissionRole.MAG_CLAMP);
+            } else {
+                baSubs.get(i).getRoles().add(MissionRole.MECHANIZED_BA);
+            }
+        }
+        generateLance(baSubs);
+        
+        return baseUnitList;
+	}
+	
+	public void generateLance(List<ForceDescriptor> subs) {
```

Comment:
Self-explanatory, but it's a large method that does a lot, so we may want some documentation on it.